File: Sources/LoeBalance/Refresh/BalanceService.swift
import Foundation

actor BalanceService {
    private let api: any APIClientProtocol
    private let auth: AuthManager
    private let snapshotStore: any SnapshotStoreProtocol
    private let reconciler: BalanceReconciler
    private let now: @Sendable () -> Date
    private let usagePageSize: Int

    private var refreshTask: Task<RefreshResult, Error>?

    init(
        api: any APIClientProtocol,
        auth: AuthManager,
        snapshotStore: any SnapshotStoreProtocol,
        reconciler: BalanceReconciler = BalanceReconciler(),
        now: @escaping @Sendable () -> Date = { Date() },
        usagePageSize: Int = 100
    ) {
        self.api = api
        self.auth = auth
        self.snapshotStore = snapshotStore
        self.reconciler = reconciler
        self.now = now
        self.usagePageSize = usagePageSize
    }

    func refresh() async throws -> RefreshResult {
        if let refreshTask {
            return try await refreshTask.value
        }

        let task = Task { [self] in
            try await performRefresh()
        }
        refreshTask = task
        defer { refreshTask = nil }
        return try await task.value
    }

    func clearBaseline() async throws {
        try snapshotStore.save(.empty)
    }

    private func performRefresh() async throws -> RefreshResult {
        let persisted = try snapshotStore.load() ?? .empty

        return try await auth.withAccessToken { [api, usagePageSize] token in
            async let currentUser = api.fetchCurrentUser(accessToken: token)
            async let dashboard = api.fetchDashboardStats(accessToken: token)
            async let usage = api.fetchUsage(accessToken: token, pageSize: usagePageSize)

            let user = try await currentUser
            let stats = try? await dashboard
            guard let records = try? await usage else {
                let snapshot = BalanceSnapshot(
                    balance: user.balance,
                    todaySpend: stats?.todayActualCost,
                    todayRequests: stats?.todayRequests,
                    updatedAt: now()
                )
                return RefreshResult(snapshot: snapshot, events: [], connectionState: .online)
            }
            let unseen = records
                .filter { !persisted.recentUsageIDs.contains($0.id) }
                .sorted { ($0.createdAt, $0.id) < ($1.createdAt, $1.id) }

            let snapshot = BalanceSnapshot(
                balance: user.balance,
                todaySpend: stats?.todayActualCost,
                todayRequests: stats?.todayRequests,
                updatedAt: now()
            )
            let events: [BalanceAnimationEvent]
            if let previous = persisted.cachedSnapshot {
                events = reconciler.reconcile(
                    previousBalance: previous.balance,
                    currentBalance: snapshot.balance,
                    unseenUsage: unseen
                )
            } else {
                events = []
            }

            var nextState = PersistedSnapshotState(
                cachedSnapshot: snapshot,
                watermarkTime: records.map(\.createdAt).max() ?? persisted.watermarkTime,
                recentUsageIDs: persisted.recentUsageIDs
            )
            nextState.recordUsageIDs(records.map(\.id))
            try snapshotStore.save(nextState)

            return RefreshResult(snapshot: snapshot, events: events, connectionState: .online)
        }
    }
}

File: windows/src/LoeBalance.Core/Refresh/BalanceService.cs
using LoeBalance.Core.Auth;
using LoeBalance.Core.Models;
using LoeBalance.Core.Networking;
using LoeBalance.Core.Persistence;

namespace LoeBalance.Core.Refresh;

public sealed class BalanceService
{
    private readonly IApiClient _api;
    private readonly AuthManager _auth;
    private readonly ISnapshotStore _snapshotStore;
    private readonly BalanceReconciler _reconciler;
    private readonly Func<DateTimeOffset> _now;
    private readonly int _usagePageSize;
    private readonly SemaphoreSlim _refreshLock = new(1, 1);

    public BalanceService(
        IApiClient api,
        AuthManager auth,
        ISnapshotStore snapshotStore,
        BalanceReconciler? reconciler = null,
        Func<DateTimeOffset>? now = null,
        int usagePageSize = 100)
    {
        _api = api;
        _auth = auth;
        _snapshotStore = snapshotStore;
        _reconciler = reconciler ?? new BalanceReconciler();
        _now = now ?? (() => DateTimeOffset.UtcNow);
        _usagePageSize = usagePageSize;
    }

    public async Task<RefreshResult> RefreshAsync(CancellationToken cancellationToken = default)
    {
        await _refreshLock.WaitAsync(cancellationToken);
        try
        {
            var persisted = await _snapshotStore.LoadAsync(cancellationToken) ?? PersistedSnapshotState.Empty;
            return await _auth.WithAccessTokenAsync(async token =>
            {
                var userTask = _api.FetchCurrentUserAsync(token, cancellationToken);
                var statsTask = _api.FetchDashboardStatsAsync(token, cancellationToken);
                var usageTask = _api.FetchUsageAsync(token, _usagePageSize, cancellationToken);
                var user = await userTask;
                var stats = await TryGetAsync(statsTask);
                var usage = await TryGetAsync(usageTask);
                var snapshot = new BalanceSnapshot(user.Balance, stats?.TodayActualCost, stats?.TodayRequests, _now());
                if (usage is null)
                {
                    return new RefreshResult(snapshot, Array.Empty<BalanceAnimationEvent>(), new ConnectionState.Online());
                }

                var unseen = usage
                    .Where(record => !persisted.RecentUsageIds.Contains(record.Id))
                    .OrderBy(record => record.CreatedAt)
                    .ThenBy(record => record.Id)
                    .ToArray();
                var events = persisted.CachedSnapshot is null
                    ? Array.Empty<BalanceAnimationEvent>()
                    : _reconciler.Reconcile(persisted.CachedSnapshot.Balance, snapshot.Balance, unseen);
                var recentIds = usage.Select(record => record.Id).Concat(persisted.RecentUsageIds).Distinct().TakeLast(500).ToList();
                var watermark = usage.Count == 0 ? persisted.WatermarkTime : usage.Max(record => record.CreatedAt);
                await _snapshotStore.SaveAsync(new PersistedSnapshotState(snapshot, watermark, recentIds), cancellationToken);
                return new RefreshResult(snapshot, events, new ConnectionState.Online());
            }, cancellationToken);
        }
        finally
        {
            _refreshLock.Release();
        }
    }

    public Task ClearBaselineAsync(CancellationToken cancellationToken = default)
        => _snapshotStore.ClearAsync(cancellationToken);

    private static async Task<T?> TryGetAsync<T>(Task<T> task)
    {
        try { return await task; }
        catch { return default; }
    }
}
